recirq/qcqmc/blueprint.py

```diff
@@ -0,0 +1,318 @@
+    params: BlueprintParams, *, base_circuit: cirq.AbstractCircuit
+) -> BlueprintData:
+    """Builds a BlueprintData from BlueprintParams.
+
+    Args:
+        params: The experiment blueprint parameters.
+        base_circuit: The circuit to shadow tomographize.
+
+    Returns:
+        A constructed BlueprintData object.
+    """
+    resolvers = list(
+        _get_resolvers(params.n_cliffords, params.qubit_partition, params.seed)
+    )
+
+    parameterized_clifford_ops: Iterable[cirq.OP_TREE] = (
+        quaff.get_parameterized_truncated_cliffords_ops(params.qubit_partition)
+    )
+
+    parameterized_clifford_circuits = tuple(
+        cirq.expand_composite(cirq.Circuit(ops), no_decomp=for_refactor.is_expected_elementary_cirq_op)
+        for ops in parameterized_clifford_ops
+    )
+    parameterized_clifford_circuit = sum(
+        parameterized_clifford_circuits, cirq.Circuit()
+    )
+
+    compiled_circuit = cirq.Circuit([base_circuit, parameterized_clifford_circuit])
+
+    circuit_with_measurement = compiled_circuit + cirq.Circuit(
+        cirq.measure(*params.qubits, key="all")
+    )
+
+    apply_optimizer_suite = {0: apply_optimizer_suite_0}[params.optimizer_suite]
+
+    optimized_circuit = apply_optimizer_suite(circuit_with_measurement)
+
+    return BlueprintData(
+        params=params,
+        compiled_circuit=optimized_circuit,
+        parameterized_clifford_circuits=parameterized_clifford_circuits,
+        resolvers=resolvers,  # type: ignore
+    )
+
+
+@attrs.frozen(repr=False)
+class BlueprintParamsRobustShadow(data.Params):
+    """Class for storing the parameters that specify a BlueprintData.
+
+    Args:
+        n_cliffords: The number of random cliffords to use during shadow tomography.
+        qubit_partition: For shadow tomography, we partition the qubits into these
+            disjoint partitions. For example, we can partition into single-qubit partitions
+            and sample from random single-qubit cliffords or put all qubits in one partition
+            and sample from random n-qubit cliffords.
+        seed: A random number seed.
+        optimizer_suite: The optimizer suite to use.
+    """
+
+    name: str
+    n_cliffords: int
+    qubit_partition: Tuple[Tuple[cirq.Qid, ...], ...]
+    seed: int = 0
+    optimizer_suite: int = 0
+
+    def __post_init__(self):
+        """A little helper to ensure that tuples end up as tuples after loading."""
+        object.__setattr__(
+            self,
+            "qubit_partition",
+            tuple(tuple(inner for inner in thing) for thing in self.qubit_partition),
+        )
+
+    @property
+    def path_string(self) -> str:
+        return config.OUTDIRS.DEFAULT_BLUEPRINT_DIRECTORY + self.name
+
+    @property
+    def qubits(self) -> Tuple[cirq.Qid, ...]:
+        """The cirq qubits."""
+        return tuple(itertools.chain(*self.qubit_partition))
+
+    def _json_dict_(self):
+        return attrs.asdict(self)
+
+
+def build_blueprint(
```

Comment:
Should this be a classmethod on BlueprintData?  such as BlueprintData.build_from_dependencies(..)

Reply:
Maybe, this pattern of the factory methods being standalone methods separate from the classes is used elsewhere in the code too but I'm not entirely sure why...

Reply:
Done.